Python/ceval.c

```diff
@@ -1929,14 +1929,73 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
         }
 
         TARGET(BINARY_MULTIPLY): {
+            PREDICTED(BINARY_MULTIPLY);
+            STAT_INC(BINARY_MULTIPLY, unquickened);
             PyObject *right = POP();
             PyObject *left = TOP();
             PyObject *res = PyNumber_Multiply(left, right);
             Py_DECREF(left);
             Py_DECREF(right);
             SET_TOP(res);
-            if (res == NULL)
+            if (res == NULL) {
                 goto error;
+            }
+            DISPATCH();
+        }
+
+        TARGET(BINARY_MULTIPLY_ADAPTIVE): {
+            if (oparg == 0) {
+                PyObject *left = SECOND();
+                PyObject *right = TOP();
+                next_instr--;
+                if (_Py_Specialize_BinaryMultiply(left, right, next_instr) < 0) {
+                    goto error;
+                }
+                DISPATCH();
+            }
+            else {
+                STAT_INC(BINARY_MULTIPLY, deferred);
+                UPDATE_PREV_INSTR_OPARG(next_instr, oparg - 1);
+                STAT_DEC(BINARY_MULTIPLY, unquickened);
+                JUMP_TO_INSTRUCTION(BINARY_MULTIPLY);
+            }
+        }
+
+        TARGET(BINARY_MULTIPLY_INT): {
+            PyObject *left = SECOND();
+            PyObject *right = TOP();
+            DEOPT_IF(!PyLong_CheckExact(left), BINARY_MULTIPLY);
+            DEOPT_IF(Py_TYPE(right) != Py_TYPE(left), BINARY_MULTIPLY);
```

Comment:
I personally prefer using `Py_IS_TYPE` instead of `==`. Currently they're equivalent, but in the future how types work *may* be changed by the subinterpreters people. When that happens they just need to change `Py_IS_TYPE`'s implementation.

```suggestion
            DEOPT_IF(Py_IS_TYPE(right, Py_TYPE(left)), BINARY_MULTIPLY);
```

Reply:
I think it's even clearer with PyLong_CheckExact twice :)

Reply:
`Py_TYPE(right) != Py_TYPE(left)` is fine for now.
`Py_IS_TYPE(right, Py_TYPE(left))` obscures the meaning, IMO.

This is the interpreter, not a third-party module. Portability is not an issue.

Reply:
> I think it's even clearer with PyLong_CheckExact twice :)

That also works :slightly_smiling_face: